### PR TITLE
switching input_metadata to using SymInts; take 2

### PR DESCRIPTION
**Pull request by Krovatkin**
Fixes #ISSUE_NUMBER

### CODE REVIEW

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -5149,6 +5157,8 @@
 
 - func: zeros(int[] size, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor
 
+- func: zeros.SymInt(SymInt[] size, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor
```
Comment (bdhirsh) [inline aten/src/ATen/native/native_functions.yaml:5160]: totally not relevant to this PR but quick question - at some point, do we wanna consider using codegen to automatically generate / enforce that all ops taking in concrete ints have a symint alternative? Maybe the infra isn't in a complete enough state that it's worth pursuing that yet?

Comment (Krovatkin) [reply aten/src/ATen/native/native_functions.yaml:5160]: I think it's pretty good idea, but we are still pretty early to pursue this yet :(

File: c10/core/SymIntArrayRef.h [hunk 2]
```diff
@@ -190,6 +190,8 @@ class SymIntArrayRef final {
 TORCH_API at::IntArrayRef asIntArrayRefSlow(c10::SymIntArrayRef ar);
 TORCH_API at::IntArrayRef asIntArrayRefUnchecked(c10::SymIntArrayRef ar);
 
-std::ostream& operator<<(std::ostream& out, const c10::SymIntArrayRef& list);
+inline std::ostream& operator<<(std::ostream& out, const c10::SymIntArrayRef& list) {
+  return out << list.wrapped_symint_array_ref;
+}
```
Comment (ezyang) [inline c10/core/SymIntArrayRef.h:195]: why is this inline now

File: aten/src/ATen/native/ReduceOps.cpp [hunk 3]
```diff
@@ -1077,6 +1079,38 @@ Tensor sum(const Tensor &self, c10::optional<ScalarType> dtype) {
...
+    ss << " but expected shape compatible with ";
+    if (ten_size.has_value()) {
+      ss << *ten_size;
+    } else {
+      ss << int_sizes;
+    }
+    // TODO: We still need to add the error prologue
+    // to match the error message exactly
+    AT_ERROR(ss.str());
+  }
```
Comment (ezyang) [inline aten/src/ATen/native/ReduceOps.cpp:1109]: can we just refactor the old error message functioninto aten

File: torch/csrc/autograd/input_metadata.h [hunk 4]
```diff
@@ -81,55 +86,18 @@ struct InputMetadata {
...
-  bool is_same_shape(const at::Tensor& grad) const {
-    TORCH_CHECK(
-        grad.is_nested() == is_nested_tensor(),
-        "Both grad and InputMetadata need to be either nested or non nested tensors.")
-    if (grad.is_nested()) {
-      return at::native::get_nested_size_tensor(grad).is_same_size(
-          shape_as_tensor());
+  c10::optional<c10::SymIntArrayRef> shape_as_dim_vector_opt() const {
+    if (is_nested_tensor()) {
+      return c10::nullopt;
```
Comment (ezyang) [inline torch/csrc/autograd/input_metadata.h:94]: lol (this is not worse than what we had before so fine lol)

**Review by ezyang — APPROVED** [review]

File: aten/src/ATen/native/native_functions.yaml [hunk 5]
```diff
@@ -4596,6 +4596,14 @@
   device_check: NoCheck
   device_guard: False
 
+- func: sum_to_checked(Tensor self, SymInt[]? size, Tensor? tsize, int grad_idx) -> Tensor
```
Comment (albanD) [inline aten/src/ATen/native/native_functions.yaml:4599]: That function is very weird, please make this private.

File: aten/src/ATen/native/ReduceOps.cpp [hunk 6]
```diff
@@ -1077,6 +1079,38 @@ Tensor sum(const Tensor &self, c10::optional<ScalarType> dtype) {
...
+  auto int_sizes = c10::asIntArrayRefSlow(*size);
+
+  if (grad.sizes().equals(int_sizes)) {
+    return grad;
+  }
+
+  if (!at::is_expandable_to(int_sizes, grad.sizes())) {
+    std::stringstream ss;
+    ss << "invalid gradient at index " << grad_idx << " - got ";
+    if (grad.is_nested()) {
```
Comment (albanD) [inline aten/src/ATen/native/ReduceOps.cpp:1095]: There is an assert above that says that this is never possible right?

**Review by albanD — CHANGES_REQUESTED** [review]
The approach sounds ok But I'm not sure the kernels are correct right now and this needs testing

**Pull request closed without merging**